src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/FileBasedProgramsProjectSystem.cs

```diff
@@ -0,0 +1,174 @@
+        {
+            // For now, we cannot provide intellisense etc on files which are not on disk or are not C#.
+            var sourceTextLoader = new SourceTextLoader(documentText, documentPath);
+            var projectInfo = MiscellaneousFileUtilities.CreateMiscellaneousProjectInfoForDocument(
+                Workspace, documentPath, sourceTextLoader, languageInformation, documentText.ChecksumAlgorithm, Workspace.CurrentSolution.Services, []);
+            await ProjectFactory.ApplyChangeToWorkspaceAsync(ws => ws.OnProjectAdded(projectInfo), cancellationToken: default);
+
+            var newSolution = Workspace.CurrentSolution;
+            if (languageInformation.LanguageName == "Razor")
+            {
+                var docId = projectInfo.AdditionalDocuments.Single().Id;
+                return newSolution.GetRequiredAdditionalDocument(docId);
+            }
+
+            var id = projectInfo.Documents.Single().Id;
+            return newSolution.GetRequiredDocument(id);
+        }
+
+        // We have a file on disk. Light up the file-based program experience.
+        var projectLanguageName = LanguageNames.CSharp;
+        var documentFileInfo = new DocumentFileInfo(documentPath, logicalPath: documentPath, isLinked: false, isGenerated: false, folders: default);
+        var projectFileInfo = new ProjectFileInfo()
+        {
+            Language = projectLanguageName,
+            FilePath = VirtualProject.GetVirtualProjectPath(documentPath),
+            CommandLineArgs = ["/langversion:preview", "/features:FileBasedProgram=true"],
+            Documents = [documentFileInfo],
+            AdditionalDocuments = [],
+            AnalyzerConfigDocuments = [],
+            ProjectReferences = [],
+            PackageReferences = [],
+            ProjectCapabilities = [],
+            ContentFilePaths = [],
+            FileGlobs = []
+        };
+
+        var projectSet = AddLoadedProjectSet(documentPath);
+        Project workspaceProject;
+        using (await projectSet.Semaphore.DisposableWaitAsync())
+        {
+            var loadedProject = await this.CreateAndTrackInitialProject_NoLockAsync(projectSet, documentPath, language: projectLanguageName);
+            await loadedProject.UpdateWithNewProjectInfoAsync(projectFileInfo, hasAllInformation: false, _logger);
+
+            ProjectsToLoadAndReload.AddWork(new ProjectToLoad(documentPath, ProjectGuid: null, ReportTelemetry: true));
+            loadedProject.NeedsReload += (_, _) => ProjectsToLoadAndReload.AddWork(new ProjectToLoad(documentPath, ProjectGuid: null, ReportTelemetry: false));
+            workspaceProject = ProjectFactory.Workspace.CurrentSolution.GetRequiredProject(loadedProject.ProjectId);
+        }
+
+        var document = workspaceProject.Documents.Single();
+        Contract.ThrowIfFalse(document.FilePath == documentPath);
+        return document;
+    }
```

Comment:
my core feedback is that this method is humongously complex, and doesn't tell a story at all as to why it's written as it is.  i really can't reason through things like:

1. why is the order the way it is?
2. why are certain bits of code here, versus in code thsi calls out to.
3. is there stuff missing?
4. what's the locking/mutation story?
5. How do we ensure all/nothing impact on our in memory state.